File: tests/oracle/jumpdest_instsym_1204.cc
/*
 * CSPEC-JUMPDEST-INSTSYM-0001: locked Ghidra 12.0.4 JUMPSYM snippet-compiler
 * oracle.  Boots a bare SLEIGH engine on the production x86-64.sla (no
 * BfdArchitecture, matching PcodeInjectLibrarySleigh::parseInject's use of
 * `const SleighBase *slgh`), then compiles p-code snippets that use the
 * language's JUMPSYM symbols through the same PcodeSnippet compiler the
 * callfixup/jumpassist ingest path uses (inject_sleigh.cc:387
 * `PcodeSnippet compiler(slgh)`), printing:
 *   - the symbol_type of each JUMPSYM name as resolved by
 *     SleighBase::findSymbol (pcodeparse.cc:3223 via PcodeSnippet::lex),
 *     proving inst_start/inst_next/inst_next2 are serialized into the
 *     .sla (slgh_compile.cc:1986-1991 predefinedSymbols) while
 *     inst_dest/inst_ref only exist in the snippet-local tree
 *     (pcodeparse.y:693-694),
 *   - the compiled ConstructTpl XML (ConstructTpl::encode(encoder,-1),
 *     the exact encoding InjectPayloadSleigh::printTemplate uses) for
 *     each snippet, covering the two grammar positions with different
 *     re-wrap rules: `jumpdest: JUMPSYM` (pcodeparse.y:195, space
 *     j_curspace + size j_curspace_size) and `varnode:
 *     specificsymbol->getVarnode()` (pcodeparse.y:202, constant space +
 *     sz_zero, then propagateSize fills from the 8-byte destination),
 *   - the lexer fallback observations: an unknown name in jumpdest
 *     position and the predefined-but-not-JUMPSYM `epsilon` symbol in
 *     varnode position.
 */
#include "loadimage.hh"
#include "marshal.hh"
#include "pcodeparse.hh"
#include "sleigh.hh"

#include <fstream>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using namespace ghidra;
using std::cerr;
using std::cout;
using std::exception;
using std::istringstream;
using std::ostringstream;
using std::runtime_error;
using std::string;
using std::vector;

// A LoadImage that answers every fetch with zeros; snippet compilation
// never fetches instruction bytes (the PcodeSnippet constructor at
// pcodeparse.y:676-696 only reads the symbol/space tables).
class ZeroLoadImage : public LoadImage {
public:
  ZeroLoadImage(void) : LoadImage("jumpdest_instsym_zero") {}
  virtual void loadFill(unsigned char *ptr, int4 size, const Address &addr)
  {
    for (int4 i = 0; i < size; ++i)
      ptr[i] = 0;
  }
  virtual string getArchType(void) const { return "jumpdest-instsym"; }
  virtual void adjustVma(long) {}
};

string escapeNewlines(const string &value)
{
  string out;
  for (string::const_iterator iter = value.begin(); iter != value.end(); ++iter) {
    if (*iter == '\n')
      out += "\\n";
    else
      out += *iter;
  }
  return out;
}

// Compile one snippet with a fresh PcodeSnippet (exactly the
// PcodeInjectLibrarySleigh::parseInject lifecycle: fresh compiler, no
// operands, default tempbase 0, one parseStream call) and print the
// result: the encoded template on success, the first error otherwise.
void runSnippet(const SleighBase *slgh, int index, const string &snippet)
{
  PcodeSnippet compiler(slgh);
  istringstream stream(snippet);
  bool ok = compiler.parseStream(stream);
  if (!ok || compiler.hasErrors()) {
    cout << "SNIP|" << index << "|ERR|" << escapeNewlines(compiler.getErrorMessage())
         << "\n";
    return;
  }
  ConstructTpl *tpl = compiler.releaseResult();
  if (tpl == (ConstructTpl *)0) {
    cout << "SNIP|" << index << "|ERR|no result\n";
    return;
  }
  ostringstream buffer;
  XmlEncode encoder(buffer);
  tpl->encode(encoder, -1);
  delete tpl;
  cout << "SNIP|" << index << "|OK|" << escapeNewlines(buffer.str()) << "\n";
}

void runFixture(const string &specDirectory)
{
  const string slaPath = specDirectory + "/x86-64.sla";
  ZeroLoadImage loader;
  ContextInternal contextDatabase;
  Sleigh engine(&loader, &contextDatabase);
  DocumentStorage store;
  // SleighArchitecture::buildSpecFile (sleigh_arch.cc:412-417): the .sla is
  // registered as a synthetic <sleigh>path</sleigh> tag; Sleigh::initialize
  // (sleigh.cc:558-568) then streams the binary file through
  // sla::FormatDecode.
  istringstream sleighTag("<sleigh>" + slaPath + "</sleigh>");
  Document *doc = store.parseDocument(sleighTag);
  store.registerTag(doc->getRoot());
  engine.initialize(store);

  cout << "SCHEMA|1\n";

  // SleighBase::findSymbol observations for every name the snippets use.
  // symbol_type ordinals follow slghsymbol.hh:28-32 (space_symbol=0 ...
  // start_symbol=9, end_symbol=10, next2_symbol=11 ... epsilon_symbol=17,
  // flowdest_symbol=19, flowref_symbol=20).
  const char *probeNames[] = {
    "inst_start", "inst_next", "inst_next2", "inst_dest", "inst_ref", "epsilon",
  };
  for (int i = 0; i < 6; ++i) {
    const SleighSymbol *sym = engine.findSymbol(probeNames[i]);
    if (sym == (const SleighSymbol *)0)
      cout << "SYM|" << probeNames[i] << "|ABSENT\n";
    else
      cout << "SYM|" << probeNames[i] << "|" << (int4)sym->getType() << "\n";
  }

  // Snippets.  Each exercises one grammar position / symbol pairing.
  const char *snippets[] = {
    "goto inst_dest;",       // 0: jumpdest x local flowdest symbol
    "goto inst_ref;",        // 1: jumpdest x local flowref symbol
    "goto inst_next;",       // 2: jumpdest x language EndSymbol
    "goto inst_start;",      // 3: jumpdest x language StartSymbol
    "goto inst_next2;",      // 4: jumpdest x language Next2Symbol
    "local x:8 = inst_ref;", // 5: varnode x local flowref (size propagation)
    "local y:8 = inst_dest;",// 6: varnode x local flowdest
    "local z:8 = inst_next;",// 7: varnode x language EndSymbol
    "goto nosuchsym;",       // 8: unknown jump destination (y:200)
    "local w:8 = epsilon;",  // 9: predefined but not JUMPSYM -> STRING
  };
  const int snippetCount = (int)(sizeof(snippets) / sizeof(snippets[0]));
  for (int i = 0; i < snippetCount; ++i)
    runSnippet((const SleighBase *)&engine, i, snippets[i]);

  cout << "DONE\n";
}

} // namespace

int main(int argc, char **argv)
{
  if (argc != 2) {
    cerr << "usage: jumpdest_instsym_1204 <specdir>\n";
    return 2;
  }
  try {
    runFixture(argv[1]);
  }
  catch (exception &error) {
    cerr << "jumpdest_instsym_1204: " << error.what() << "\n";
    return 1;
  }
  catch (...) {
    cerr << "jumpdest_instsym_1204: unknown exception\n";
    return 1;
  }
  return 0;
}

File: tests/oracle/jumpdest_instsym_1204.rs
//! CSPEC-JUMPDEST-INSTSYM-0001: Rust side of the locked Ghidra 12.0.4
//! JUMPSYM snippet-compiler oracle.  Compiles the same ten snippets as the
//! C++ fixture (tests/oracle/jumpdest_instsym_1204.cc) through
//! `PcodeSnippet` + `PredefinedJumpSymbols`, printing the same projections
//! byte for byte:
//!   - `SYM|` — for each probed name, the `symbol_type` ordinal
//!     (slghsymbol.hh:28-32) the language lookup resolves it to.  The three
//!     predefined JUMPSYM symbols are folded into the language-side wrapper
//!     (`PredefinedJumpSymbols`, standing in for the .sla-serialized
//!     StartSymbol/EndSymbol/Next2Symbol of slgh_compile.cc:1986-1991);
//!     `inst_dest`/`inst_ref`/`epsilon` are ABSENT from the language table
//!     exactly like the locked oracle (inst_dest/inst_ref live only in the
//!     snippet-local tree, pcodeparse.y:693-694; EpsilonSymbol is dropped by
//!     SymbolTable::purge, slghsymbol.cc:248-255).
//!   - `SNIP|i|OK|<xml>` / `SNIP|i|ERR|<msg>` — the compiled template
//!     through the same XmlEncode formatting the C++ fixture's
//!     `ConstructTpl::encode(encoder,-1)` produces.
//!
//! Snippet temporaries start at unique offset 0 (PcodeSnippet default
//! tempbase, pcodeparse.y:681) matching the C++ side.

use rugra::pcodeparse::{
    ConstTpl, ConstructTpl, OpTpl, PcodeSnippet, PredefinedJumpSymbols, SleighSymbol,
    SleighSymbolLookup, SleightSymbolKind, VarnodeTpl,
};

use std::process::ExitCode;
use std::sync::Arc;

/// symbol_type ordinals (slghsymbol.hh:28-32) for the wrapper-provided
/// language symbols, matching the locked oracle's `findSymbol` result.
fn symbol_type_ordinal(sym: &SleighSymbol) -> Option<i32> {
    match &sym.kind {
        // space_symbol=0, token_symbol=1, userop_symbol=2, value_symbol=3,
        // valuemap_symbol=4, name_symbol=5, varnode_symbol=6,
        // varnodelist_symbol=7, operand_symbol=8, start_symbol=9,
        // end_symbol=10, next2_symbol=11, ...
        SleightSymbolKind::Space(_) => Some(0),
        SleightSymbolKind::UserOp(_) => Some(2),
        SleightSymbolKind::Varnode(_) => Some(6),
        SleightSymbolKind::Operand(_, _) => Some(8),
        SleightSymbolKind::JumpTarget(kind) => match kind {
            rugra::pcodeparse::JumpTargetKind::InstStart => Some(9),
            rugra::pcodeparse::JumpTargetKind::InstNext => Some(10),
            rugra::pcodeparse::JumpTargetKind::InstNext2 => Some(11),
            // flowdest_symbol=19, flowref_symbol=20 (language lookup never
            // returns these: they are snippet-local only).
            rugra::pcodeparse::JumpTargetKind::InstDest => Some(19),
            rugra::pcodeparse::JumpTargetKind::InstRef => Some(20),
        },
        // epsilon_symbol=17, label_symbol=18 — not reachable through the
        // language wrapper.
        SleightSymbolKind::Label(_, _) => Some(18),
    }
}

/// The XmlEncode-style mini writer used to print compiled templates with
/// byte-identical formatting to Ghidra's `ConstructTpl::encode`
/// (marshal.cc XmlEncode).
struct TplWriter {
    out: String,
    depth: usize,
    tag_open: bool,
}

impl TplWriter {
    fn new() -> Self {
        Self {
            out: String::new(),
            depth: 0,
            tag_open: false,
        }
    }
    fn newline_indent(&mut self) {
        self.out.push('\n');
        for _ in 0..self.depth {
            self.out.push_str("  ");
        }
    }
    fn begin(&mut self, name: &str, attrs: &[(&str, String)]) {
        if self.tag_open {
            self.out.push('>');
            self.tag_open = false;
        }
        self.newline_indent();
        self.out.push('<');
        self.out.push_str(name);
        for (key, value) in attrs {
            self.out.push_str(&format!(" {}=\"{}\"", key, value));
        }
        self.depth += 1;
        self.tag_open = true;
    }
    fn end(&mut self, name: &str) {
        self.depth -= 1;
        if self.tag_open {
            self.out.push_str("/>");
            self.tag_open = false;
            return;
        }
        self.newline_indent();
        self.out.push_str("</");
        self.out.push_str(name);
        self.out.push('>');
    }
}

fn write_const_tpl(writer: &mut TplWriter, ct: &ConstTpl) {
    // ConstTpl::encode element names (semantics.cc:300-365, slaformat.cc).
    let (name, attrs): (&str, Vec<(&str, String)>) = match ct {
        ConstTpl::Real(v) => ("const_real", vec![("val", format!("0x{:x}", v))]),
        ConstTpl::SpaceId(_) => ("const_spaceid", Vec::new()), // space filled by caller
        ConstTpl::JCurSpace => ("const_curspace", Vec::new()),
        ConstTpl::JCurSpaceSize => ("const_curspace_size", Vec::new()),
        ConstTpl::JRelative(i) => ("const_relative", vec![("val", format!("0x{:x}", i))]),
        ConstTpl::JStart => ("const_start", Vec::new()),
        ConstTpl::JNext => ("const_next", Vec::new()),
        ConstTpl::JNext2 => ("const_next2", Vec::new()),
        ConstTpl::JFlowRef => ("const_flowref", Vec::new()),
        ConstTpl::JFlowDest => ("const_flowdest", Vec::new()),
        ConstTpl::Handle { index, select, plus } => {
            // ConstTpl::encode handle case (semantics.cc:309-316): val =
            // handle index, s = select, plus only for v_offset_plus.
            let s = match select {
                rugra::pcodeparse::HandleSelect::Space => 0,
                rugra::pcodeparse::HandleSelect::Offset => 1,
                rugra::pcodeparse::HandleSelect::Size => 2,
                rugra::pcodeparse::HandleSelect::OffsetPlus => 3,
            };
            if matches!(select, rugra::pcodeparse::HandleSelect::OffsetPlus) {
                (
                    "const_handle",
                    vec![
                        ("val", format!("{}", index)),
                        ("s", format!("{}", s)),
                        ("plus", format!("0x{:x}", plus)),
                    ],
                )
            } else {
                ("const_handle", vec![("val", format!("{}", index)), ("s", format!("{}", s))])
            }
        }
    };
    if name == "const_spaceid" {
        // writeSpace emits the space name (locked x86-64 names).
        let spc = match ct {
            ConstTpl::SpaceId(s) => *s,
            _ => unreachable!(),
        };
        let spc_name = match spc {
            rugra::space::AddressSpace::Const => "const",
            rugra::space::AddressSpace::Unique => "unique",
            rugra::space::AddressSpace::Ram => "ram",
            rugra::space::AddressSpace::Register => "register",
            rugra::space::AddressSpace::Stack => "stack",
            rugra::space::AddressSpace::Iop => "iop",
            rugra::space::AddressSpace::Join => "join",
            rugra::space::AddressSpace::Other(_) => "OTHER",
            rugra::space::AddressSpace::Overlay => "OTHER",
        };
        writer.begin(name, &[("space", spc_name.to_string())]);
    } else {
        writer.begin(name, &attrs);
    }
    writer.end(name);
}

fn write_varnode_tpl(writer: &mut TplWriter, vn: &VarnodeTpl) {
    writer.begin("varnode_tpl", &[]);
    write_const_tpl(writer, &vn.get_space());
    write_const_tpl(writer, &vn.get_offset());
    write_const_tpl(writer, &vn.get_size());
    writer.end("varnode_tpl");
}

/// Byte-identical mirror of the C++ fixture's `tpl->encode(encoder, -1)`
/// (ConstructTpl::encode, semantics.cc: the <construct_tpl> header with a
/// <null/> result handle then one <op_tpl> per op).
fn write_template(tpl: &ConstructTpl) -> String {
    let mut writer = TplWriter::new();
    writer.begin("construct_tpl", &[]);
    // Snippet templates carry no result handle: <null/>.
    writer.begin("null", &[]);
    writer.end("null");
    for op in tpl.get_opvec() {
        write_op_tpl(&mut writer, op);
    }
    writer.end("construct_tpl");
    writer.out
}

fn write_op_tpl(writer: &mut TplWriter, op: &OpTpl) {
    writer.begin("op_tpl", &[("code", op.opc.name().to_string())]);
    match &op.out {
        Some(out) => write_varnode_tpl(writer, out),
        None => {
            writer.begin("null", &[]);
            writer.end("null");
        }
    }
    for input in &op.inputs {
        write_varnode_tpl(writer, input);
    }
    writer.end("op_tpl");
}

fn escape_newlines(value: &str) -> String {
    value.replace('\n', "\\n")
}

/// The language-side host: only the three predefined JUMPSYM symbols
/// (everything else resolves through the inner empty lookup, i.e. ABSENT,
/// exactly like the locked x86-64.sla for inst_dest/inst_ref/epsilon).
struct EmptyLanguage;

impl SleighSymbolLookup for EmptyLanguage {
    fn find_symbol(&self, _name: &str) -> Option<SleighSymbol> {
        None
    }
}

fn run_snippet(index: usize, snippet: &str, out: &mut String) {
    // parseInject lifecycle (inject_sleigh.cc:387-416): fresh compiler per
    // snippet, language lookup installed, default tempbase.
    let mut compiler = PcodeSnippet::new();
    compiler.set_sleigh_lookup(Arc::new(PredefinedJumpSymbols::new(EmptyLanguage)));
    if compiler.parse_stream(snippet) && !compiler.has_errors() {
        if let Some(tpl) = compiler.release_result() {
            out.push_str(&format!(
                "SNIP|{}|OK|{}\n",
                index,
                escape_newlines(&write_template(&tpl))
            ));
            return;
        }
        out.push_str(&format!("SNIP|{}|ERR|no result\n", index));
        return;
    }
    out.push_str(&format!(
        "SNIP|{}|ERR|{}\n",
        index,
        escape_newlines(compiler.get_error_message())
    ));
}

fn run() -> Result<(), String> {
    let language = PredefinedJumpSymbols::new(EmptyLanguage);

    let mut out = String::new();
    out.push_str("SCHEMA|1\n");

    // SleighBase::findSymbol observations.  inst_dest/inst_ref/epsilon are
    // absent from the language table (the C++ oracle prints ABSENT; the
    // local snippet tree provides inst_dest/inst_ref separately).
    let probe_names = ["inst_start", "inst_next", "inst_next2", "inst_dest", "inst_ref", "epsilon"];
    for name in probe_names {
        match language.find_symbol(name) {
            None => out.push_str(&format!("SYM|{}|ABSENT\n", name)),
            Some(sym) => match symbol_type_ordinal(&sym) {
                Some(ordinal) => out.push_str(&format!("SYM|{}|{}\n", name, ordinal)),
                None => return Err(format!("symbol {} has no type ordinal", name)),
            },
        }
    }

    // The same ten snippets as the C++ fixture.
    let snippets = [
        "goto inst_dest;",        // 0: jumpdest x local flowdest symbol
        "goto inst_ref;",         // 1: jumpdest x local flowref symbol
        "goto inst_next;",        // 2: jumpdest x language EndSymbol
        "goto inst_start;",       // 3: jumpdest x language StartSymbol
        "goto inst_next2;",       // 4: jumpdest x language Next2Symbol
        "local x:8 = inst_ref;",  // 5: varnode x local flowref (size propagation)
        "local y:8 = inst_dest;", // 6: varnode x local flowdest
        "local z:8 = inst_next;", // 7: varnode x language EndSymbol
        "goto nosuchsym;",        // 8: unknown jump destination (y:200)
        "local w:8 = epsilon;",   // 9: predefined-but-purged name -> STRING
    ];
    for (index, snippet) in snippets.iter().enumerate() {
        run_snippet(index, snippet, &mut out);
    }

    out.push_str("DONE\n");
    print!("{}", out);
    Ok(())
}

fn main() -> ExitCode {
    match run() {
        Ok(()) => ExitCode::SUCCESS,
        Err(message) => {
            eprintln!("jumpdest_instsym_1204: {}", message);
            ExitCode::FAILURE
        }
    }
}
